### PR TITLE
oauth: preserve query parameters in the authorization_endpoint

### PR DESCRIPTION
**Pull request by qfiard**
Query parameters are needed to customize the authentication flow for some providers, e.g. for Google Oauth2
- access_type=offline is needed to generate both an access and refresh token
- prompt=select_account is needed to ask the user to select an account
https://developers.google.com/identity/protocols/oauth2/openid-connect#authenticationuriparameters

This PR is forked from https://github.com/envoyproxy/envoy/pull/18650 as I could not find a way to re-open it. If you are able to re-open https://github.com/envoyproxy/envoy/pull/18650, I'll switch to using so we can keep the discussion in one place.

Signed-off-by: Quentin Fiard <quentin.fiard@gmail.com>

### CODE REVIEW

**Comment by wbpcode** [comment]
is it ok for you to review this PR  @snowp
I know you are very busy. So feel free to assign to me if you have no free time and I will do some pre-review.

**Review by snowp — CHANGES_REQUESTED** [review]
Makes sense to me, a few comments

File: source/extensions/filters/http/oauth2/filter.h [hunk 1]
```diff
@@ -134,14 +134,24 @@ class FilterConfig {
...
-  const std::string& authorizationEndpoint() const { return authorization_endpoint_; }
+  absl::string_view authorizationEndpointPathAndQueryParams() const {
+    return authorization_endpoint_url_.pathAndQueryParams();
+  }
+  absl::string_view authorizationEndpointHostAndPort() const {
+    return authorization_endpoint_url_.hostAndPort();
+  }
+  absl::string_view authorizationEndpointScheme() const {
+    return authorization_endpoint_url_.scheme();
+  }
```
[the remark this answers is not in the document]
Comment (qfiard) [reply source/extensions/filters/http/oauth2/filter.h]: Done.

File: source/extensions/filters/http/oauth2/filter.cc [hunk 2]
```diff
@@ -315,11 +326,16 @@ Http::FilterHeadersStatus OAuth2Filter::decodeHeaders(Http::RequestHeaderMap& he
...
-        AuthorizationEndpointFormat, config_->authorizationEndpoint(), config_->clientId(),
-        config_->encodedAuthScopes(), escaped_redirect_uri, escaped_state);
+    auto query_params = config_->authorizationQueryParams();
+    query_params["redirect_uri"] = escaped_redirect_uri;
+    query_params["state"] = escaped_state;
+    const std::string path_and_query_params = Http::Utility::replaceQueryString(
+        Http::HeaderString(config_->authorizationEndpointPathAndQueryParams()), query_params);
+    const std::string new_url = absl::StrCat(
+        config_->authorizationEndpointScheme(), "://", config_->authorizationEndpointHostAndPort(),
+        path_and_query_params, config_->encodedResourceQueryParams());
```
[the remark this answers is not in the document]
Comment (qfiard) [reply source/extensions/filters/http/oauth2/filter.cc]: Done. Added a toString() method on Url and a setPathAndQueryParams so this logic can live in Http::Utility::Url. Please let me know if you have any concerns.

**Comment by qfiard** [comment]
@snowp This pull request should be ready for another pass, thanks for the review.

File: test/common/http/utility_test.cc [hunk 3]
```diff
@@ -1495,12 +1495,18 @@ TEST(Url, ParsingFails) {
...
   Utility::Url url;
   ASSERT_TRUE(url.initialize(raw_url, false)) << "Failed to initialize " << raw_url;
   EXPECT_EQ(url.scheme(), expected_scheme);
   EXPECT_EQ(url.hostAndPort(), expected_host_port);
   EXPECT_EQ(url.pathAndQueryParams(), expected_path);
+  if (validate_tostring) {
+    // Verify that url.toString() is equivalent to raw_url.
+    validateUrl(url.toString(), expected_scheme, expected_host_port, expected_path,
+                /*validate_tostring=*/false);
+  }
```
Comment (snowp) [inline test/common/http/utility_test.cc:1509]: This seems a bit hard to grok, seems like we're just testing that url.toString() does the right thing after parsing a string? I'm not sure if this is worth testing, but if there is a good reason I'd try to refactor this a bit to make it easier to understand what's going on

**Review by snowp — CHANGES_REQUESTED** [review]
Sorry for the review delay, this looks good modulo the one comment

**Comment by qfiard** [comment]
Same here, please let me know if I can help in any way to get this merged.

File: test/common/http/utility_test.cc (hunk 3 above)
Comment (snowp) [reply test/common/http/utility_test.cc:1530]: Ping on this

**Comment by snowp** [comment]
Waiting on https://github.com/envoyproxy/envoy/pull/23501#discussion_r1046001665

/wait

File: test/common/http/utility_test.cc (hunk 3 above)
Comment (qfiard) [reply test/common/http/utility_test.cc]: Reverted these changes, toString() is simple enough that it should be fine not to have dedicated tests for it.

**Comment by qfiard** [comment]
Thanks for the review and apologies for missing https://github.com/envoyproxy/envoy/pull/23501#discussion_r1046001665. I have reverted the tests for toString() which should be simple enough not to need tests to verify that it works for all types of URLs.

**Review by snowp — APPROVED** [review]
Thanks!

**Pull request merged**